src/polyswarm_api/__init__.py

```diff
@@ -310,67 +313,84 @@ def _fix_result(self, result):
+        futures = []
+        file_objs = grouper(MAX_ARTIFACT_BATCH_SIZE, file_objs)
+        file_names = grouper(MAX_ARTIFACT_BATCH_SIZE, file_names)
+        for artifacts, names in zip(file_objs, file_names):
+            futures.append(self.post_artifacts(artifacts, names, artifact_type=artifact_type))
+        return await asyncio.gather(*futures)
 
-        return await self._wait_for_uuid(uuid)
+    async def wait_for_results(self, results):
+        futures = []
+        for result in results:
+            futures.append(self._wait_for_uuid(result['result']))
+        return await asyncio.gather(*futures)
 
-    async def scan_fileobj(self, to_scan, filename='data'):
+    async def scan_urls(self, to_scan):
         """
-        Scan a single file-like object using the PS API asynchronously.
+        Scan a single URL using the PS API asynchronously
 
-        :param to_scan: File-like object to scan.
-        :param filename: Filename to use
+        :param to_scan: List of URLs to scan
         :return: JSON report
         """
+        try:
+            file_objs = [io.StringIO(url) for url in to_scan]
+            file_names = ['url'] * len(to_scan)
+            result = await self.scan_fileobjs(file_objs, file_names, artifact_type=ArtifactType.URL)
+            return await self.wait_for_results(result)
+        except exceptions.PolyswarmAPIException as e:
+            return {'filename': str(e), 'files': []}
 
-        result = await self.post_file(to_scan, filename)
-        if result['status'] == 'OK':
-            uuid = result['result']
-        else:
-            logger.error('Failed to get UUID for scan of file %s', filename)
-            return {'filename': filename, 'files': []}
-
-        logger.info('Successfully submitted file %s, UUID %s', filename, uuid)
-
-        return await self._wait_for_uuid(uuid)
-
-    async def scan_data(self, data, filename=None):
+    async def scan_files(self, to_scan):
         """
-        Scan bytes using the PS API asynchronously.
+        Scan a collection of files using the PS API asynchronously.
 
-        :param data: Data (in bytes) to submit to be scanned
-        :param filename: Filename to use in submission
-        :return: JSON report
+        :param to_scan: List of paths of files to scan.
+        :return: JSON report file
         """
-        if filename is None:
-            filename = hashlib.sha256(data).hexdigest()
-        return await self.scan_fileobj(io.BytesIO(data), filename)
+        # early definition to avoid exceptions in try..catch in the finally clause
+        file_objs = []
+        try:
+            file_objs = [open(file_name, 'rb') for file_name in to_scan]
+            file_names = [os.path.basename(file_name) for file_name in to_scan]
+            result = await self.scan_fileobjs(file_objs, file_names, artifact_type=ArtifactType.FILE)
+            return await self.wait_for_results(result)
+        except exceptions.PolyswarmAPIException as e:
```

Comment:
Same comment as above, add in `status` and `result` and move handling to the `post_artifacts` function.

Reply:
Fixed :+1: